File: web/src/show.ts
import { existsSync } from "node:fs";
import { join } from "node:path";

import { getCurrentCommit } from "./branch.ts";
import {
	extractContentFromBlob,
	extractTreeFromCommit,
	parseTreeEntries,
	readObject,
	type PackfileCache,
} from "./utils.ts";

export async function show(path: string, filePath: string): Promise<string> {
	const gitDir = join(path, ".git");

	if (!existsSync(gitDir)) {
		throw new Error("Not a git repository");
	}

	const commitSha = await getCurrentCommit(gitDir);
	if (commitSha === null) {
		throw new Error(`Path '${filePath}' does not exist in 'HEAD'`);
	}

	const cache: PackfileCache = new Map();
	const commitData = await readObject(gitDir, commitSha, cache);
	const treeSha = extractTreeFromCommit(commitData);

	const blobSha = await resolveBlobSha(gitDir, treeSha, filePath, cache);
	if (blobSha === null) {
		throw new Error(`Path '${filePath}' does not exist in 'HEAD'`);
	}

	const blobData = await readObject(gitDir, blobSha, cache);
	return extractContentFromBlob(blobData);
}

async function resolveBlobSha(
	gitDir: string,
	treeSha: string,
	filePath: string,
	cache: PackfileCache,
): Promise<string | null> {
	const parts = filePath.split("/");
	let currentSha = treeSha;

	for (let i = 0; i < parts.length; i++) {
		const isLast = i === parts.length - 1;
		const treeData = await readObject(gitDir, currentSha, cache);
		const entries = parseTreeEntries(treeData);
		const entry = entries.find((e) => e.path === parts[i]);
		if (!entry) {
			return null;
		}
		if (isLast) {
			return entry.type === "blob" ? entry.sha : null;
		}
		if (entry.type !== "tree") {
			return null;
		}
		currentSha = entry.sha;
	}

	return null;
}

File: dotnet/Gitologist/Show.cs
namespace Gitologist;

public static class Show
{
    public static async Task<string> ShowFile(string path, string filePath, string? commit = null)
    {
        var gitDir = Path.Combine(path, ".git");

        if (!Directory.Exists(gitDir))
        {
            throw new InvalidOperationException("Not a git repository");
        }

        string commitSha;
        if (commit != null)
        {
            commitSha = commit;
        }
        else
        {
            var headSha = await Branch.GetCurrentCommit(gitDir);
            if (headSha == null)
            {
                throw new InvalidOperationException($"Path '{filePath}' does not exist in 'HEAD'");
            }
            commitSha = headSha;
        }

        var cache = new Utils.PackfileCache();

        string commitData;
        if (commit != null)
        {
            string data;
            try
            {
                data = await Utils.ReadObject(gitDir, commitSha, cache);
            }
            catch
            {
                throw new InvalidOperationException($"Commit '{commitSha}' not found");
            }
            if (!data.StartsWith("commit "))
            {
                throw new InvalidOperationException($"Commit '{commitSha}' not found");
            }
            commitData = data;
        }
        else
        {
            commitData = await Utils.ReadObject(gitDir, commitSha, cache);
        }

        var treeSha = Utils.ExtractTreeFromCommit(commitData);

        var blobSha = await ResolveBlobSha(gitDir, treeSha, filePath, cache);
        if (blobSha == null)
        {
            throw new InvalidOperationException($"Path '{filePath}' does not exist in 'HEAD'");
        }

        var blobData = await Utils.ReadObject(gitDir, blobSha, cache);
        return Utils.ExtractContentFromBlob(blobData);
    }

    private static async Task<string?> ResolveBlobSha(
        string gitDir,
        string treeSha,
        string filePath,
        Utils.PackfileCache cache
    )
    {
        var parts = filePath.Split('/');
        var currentSha = treeSha;

        for (int i = 0; i < parts.Length; i++)
        {
            var isLast = i == parts.Length - 1;
            var treeData = await Utils.ReadObject(gitDir, currentSha, cache);
            var entries = Utils.ParseTreeEntries(treeData);

            var entry = entries.FirstOrDefault(e => e.Path == parts[i]);
            if (entry == null)
            {
                return null;
            }

            if (isLast)
            {
                return entry.Type == "blob" ? entry.Sha : null;
            }

            if (entry.Type != "tree")
            {
                return null;
            }

            currentSha = entry.Sha;
        }

        return null;
    }
}
